### PR TITLE
Allowed more processes to await on an event and get triggered.

### PR DESCRIPTION
**Pull request by aromanro**
Also removed a naked pointer.

### CODE REVIEW

File: include/fschuetz04/simcpp20/simulation.hpp [hunk 1]
```diff
@@ -42,17 +43,17 @@ template <typename Time = double> class simulation {
    * @return New pending value event.
    */
   template <typename Value> value_event<Value, Time> event() {
-    return value_event<Value, Time>{*this};
+	return value_event<Value, Time>{*this};
```
Comment (fschuetz04) [inline include/fschuetz04/simcpp20/simulation.hpp:46]: I don't want to include the formatting changes.

**Comment by fschuetz04** [comment]
With this PR, aborting processes is no longer possible right? Do you think that that is an unnecessary feature?

File: include/fschuetz04/simcpp20/simulation.hpp (hunk 1 above)
Comment (aromanro) [reply include/fschuetz04/simcpp20/simulation.hpp:46]: I did not use anything :)
When I remember, I format with what's included in Visual Studio.

**Comment by aromanro** [comment]
> With this PR, aborting processes is no longer possible, right? Do you think that that is an unnecessary feature?

In my example I just added an ` env.Packet_event.abort();` after the 
```
 co_await env.Packet_event;
 packet = env.get_Packet();
```

in Receiver's Run. 
I've got what I expected:
```
Cable waking up but has nothing to send at 3.
Sender sending 1 packet at 5.
Cable waking up and sending packet 1 at 6.
Receiver 1 received 1 packet at 6.
Receiver 2 received 1 packet at 6.
Receiver 3 received 1 packet at 6.
Cable waking up but has nothing to send at 9.
Sender sending 2 packet at 10.
Cable waking up and sending packet 2 at 12.
Sender sending 3 packet at 15.
Cable waking up and sending packet 3 at 15.
Cable waking up but has nothing to send at 18.
Sender sending 4 packet at 20.
Cable waking up and sending packet 4 at 21.
Cable waking up but has nothing to send at 24.
Sender sending 5 packet at 25.
Cable waking up and sending packet 5 at 27.
Sender sending 6 packet at 30.
Cable waking up and sending packet 6 at 30.
Cable waking up but has nothing to send at 33.
Sender sending 7 packet at 35.
Cable waking up and sending packet 7 at 36.
Cable waking up but has nothing to send at 39.
Sender sending 8 packet at 40.
Cable waking up and sending packet 8 at 42.
Sender sending 9 packet at 45.
Cable waking up and sending packet 9 at 45.
Cable waking up but has nothing to send at 48.
Sender sending 10 packet at 50.
Cable waking up and sending packet 10 at 51.
Cable waking up but has nothing to send at 54.
Sender sending 11 packet at 55.
Cable waking up and sending packet 11 at 57.
Sender sending 12 packet at 60.
Cable waking up and sending packet 12 at 60.
Cable waking up but has nothing to send at 63.
Sender sending 13 packet at 65.
Cable waking up and sending packet 13 at 66.
Cable waking up but has nothing to send at 69.
Sender sending 14 packet at 70.
Cable waking up and sending packet 14 at 72.
Sender sending 15 packet at 75.
Cable waking up and sending packet 15 at 75.
Cable waking up but has nothing to send at 78.
Sender sending 16 packet at 80.
Cable waking up and sending packet 16 at 81.
Cable waking up but has nothing to send at 84.
Sender sending 17 packet at 85.
Cable waking up and sending packet 17 at 87.
Sender sending 18 packet at 90.
Cable waking up and sending packet 18 at 90.
Cable waking up but has nothing to send at 93.
Sender sending 19 packet at 95.
Cable waking up and sending packet 19 at 96.
Cable waking up but has nothing to send at 99.
Sender sending 20 packet at 100.
Cable waking up and sending packet 20 at 102.
```
No packets received after the first trigger.

**Comment by fschuetz04** [comment]
Take the following code:

```cppp
#include <cstdio>

#include "fschuetz04/simcpp20.hpp"

simcpp20::event<> a(simcpp20::simulation<> &sim) {
  printf("[%.0f] a 1\n", sim.now());
  co_return;
}

int main() {
  simcpp20::simulation<> sim;
  auto ev_a = a(sim);
  ev_a.abort();
  sim.run();
}
```

With this change, the line will be printed, even though the process was aborted. I don't know whether that is an important feature, but its definitely a behavior change :)

**Comment by aromanro** [comment]
I see... well, I'll check what can be done about that.

**Comment by aromanro** [comment]
Ok, now your example appears to work all right, mine also, tests pass... at least compiled with VC++.

The problem was that the 'event' on which abort() was called (the returned one by a(sim)) was not the same as the one that contained the handle.

**Comment by aromanro** [comment]
I see that even for compilers those coroutines are too complex :)

"internal compiler error: Segmentation fault"

**Comment by aromanro** [comment]
There are still issues with value_event.

**Comment by aromanro** [comment]
Ok, looks like it's not that easy, I'll close this for now.

**Comment by fschuetz04** [comment]
@aromanro I tried to implement this, unfortunately still using raw pointers, but I don't really know a better way in this case. Can you please test your example with the latest commit? Thanks!

**Comment by aromanro** [comment]
I'll look over the latest one today. I tried my example with your changes yesterday, I had to modify it in order to have it working again. 

Here is the changed version:
```
#define OBSERVABLE_PROPERTY(TYP, NAM, VAL)                                     \
private:                                                                       \
  TYP NAM = VAL;                                                               \
  simcpp20::event<> NAM##_event{sim};                                          \
																			   \
public:                                                                        \
  simcpp20::event<> get_##NAM##_event() { return this->NAM##_event; }          \
  TYP get_##NAM() {                                                            \
	return this->NAM;                                                          \
  }                                                                            \
  void set_##NAM(TYP v) {                                                      \
	this->NAM = v;                                                             \
	this->NAM##_event.trigger();                                               \
	this->NAM##_event = simcpp20::event<>(sim);                                \
  }


class EnvObj {
public:
  simcpp20::simulation<> &sim;

  explicit EnvObj(simcpp20::simulation<> &sim) : sim(sim) {}
  virtual ~EnvObj() = default;
};


class Environment : public EnvObj {
public:
  using EnvObj::EnvObj;
  ~Environment() noexcept override = default;

  OBSERVABLE_PROPERTY(int, Packet, 0);
};



class Cable {
public:
  explicit Cable(Environment& env) 
	  : sim(env.sim), env(env)
  { 
	  Run();
  }

  simcpp20::event<> Run() 
  {
	  while (env.sim.now() < 100) {
		auto timeout = env.sim.timeout(3);
		co_await timeout;

		if (!q.empty()) 
		{
		  printf("Cable waking up and sending packet %d at %g.\n", q.front(), env.sim.now());
		  
		  int packet = getPacket();
		 
		  //env.get_Packet_event().abort();
		  env.set_Packet(packet);
		} else printf("Cable waking up but has nothing to send at %g.\n", env.sim.now());
	  }
  }

  int getPacket() {
	if (q.empty())
	  return 0;

	int res = q.front();
	q.pop();
	return res;
  }

  void SendPacket(int data) { q.push(data); }
private:
  std::queue<int> q;

public:
  simcpp20::simulation<> &sim;
  Environment &env;
};

class Sender {
public:
  explicit Sender(Environment &env, Cable &cable) : sim(env.sim), cable(cable) 
  {
	Run();
  }

   simcpp20::event<> Run() {
	while (sim.now() < 100) {
	  co_await sim.timeout(5);

	  ++packet;

	  printf("Sender sending %d packet at %g.\n", packet, sim.now());
	  cable.SendPacket(packet);
	}
  }

  simcpp20::simulation<> &sim;

private:
  int packet = 0;
  Cable& cable;
};

class Receiver {
public:
  explicit Receiver(Environment &env, Cable &cable, int recvId)
	  : sim(env.sim), env(env), cable(cable), recvId(recvId) {
	Run();
  }

  simcpp20::event<> Run() {
	while (env.sim.now() < 100) {
	  co_await env.get_Packet_event();
	  int packet = env.get_Packet();

	  //env.get_Packet_event().abort();
		 
	  if (packet)
		  printf("Receiver %d received %d packet at %g.\n", recvId, packet, env.sim.now());
	  else
		  printf("Odd, empty packet received by receiver %d\n", recvId);
	}

	co_return;
  }

  simcpp20::simulation<> &sim;
  Environment &env;
  Cable &cable;
  int recvId;
};

int main() {
  simcpp20::simulation<> sim;
  Environment env(sim);

  Cable cable(env);
  Sender sender(env, cable);
  Receiver receiver1(env, cable, 1);
  Receiver receiver2(env, cable, 2);
  Receiver receiver3(env, cable, 3);

  sim.run();
}
```

I'm a little uneasy about not only raw pointers, but also on how objects are passed by value (copied, whatever)... I think I'll do my own implementation avoiding both those ugly 'protothreads' and also coroutines (at least until I'll learn more about them). There are some specific needs that I think could be better met with my custom implementation, anyway.

**Comment by fschuetz04** [comment]
I added another commit this morning, now you shouldn't need to copy the event via get_Packet_event anymore. Can you please verify? Also, value events might be useful for this kind of stuff in general.

**Comment by aromanro** [comment]
As I commented on the 'memory_leaks' issue, both examples seem to work.
One minor thing, you have this:
```
 auto temp_promises = data_->promises_;
 data_->promises_.clear();
```

You could achieve the same thing by simply swapping the content of the vector, something like this:
```
    std::vector<generic_promise_type *> temp_promises;
    data_->promises_.swap(temp_promises);
```

**Comment by fschuetz04** [comment]
Thanks, I changed it.

**Pull request closed without merging**